src/style-spec/function/expressions.js

```diff
@@ -0,0 +1,354 @@
+'use strict';
+
+// @flow
+
+const assert = require('assert');
+
+const {
+    NumberType,
+    StringType,
+    BooleanType,
+    ColorType,
+    ObjectType,
+    ValueType,
+    InterpolationType,
+    typename,
+    variant,
+    vector,
+    array,
+    anyArray,
+    lambda,
+    nargs
+} = require('./types');
+
+/*::
+ import type { PrimitiveType, TypeName, VariantType, VectorType, ArrayType, AnyArrayType, NArgs, LambdaType, Type } from './types.js';
+
+ import type { TypeError, TypedLambdaExpression, TypedLiteralExpression, TypedExpression } from './type_check.js';
+
+ import type { CompiledExpression } from './compile.js';
+
+ export type ExpressionName = "literal" | "ln2" | "pi" | "e" | "string" | "number" | "boolean" | "json_array" | "object" | "get" | "has" | "at" | "typeof" | "length" | "zoom" | "properties" | "geometry_type" | "id" | "case" | "match" | "is_error" | "==" | "!=" | ">" | ">=" | "<=" | "<" | "&&" | "||" | "!" | "curve" | "step" | "exponential" | "linear" | "cubic-bezier" | "+" | "-" | "*" | "/" | "%" | "^" | "log10" | "ln" | "log2" | "sin" | "cos" | "tan" | "asin" | "acos" | "atan" | "ceil" | "floor" | "round" | "abs" | "min" | "max" | "concat" | "upcase" | "downcase" | "rgb" | "rgba" | "color" | "color_to_array"
+
+ type Definition = {
+     name: ExpressionName,
+     type: Type,
+     compile: (expr: TypedExpression, args: Array<CompiledExpression>) => ({ js?: string, errors?: Array<string>, isFeatureConstant?: boolean, isZoomConstant?: boolean })
+ }
+ */
+
+const expressions: { [string]: Definition } = {
```

Comment:
Ack.  Seems like this isn't compatible with accessing this map using `string` values.  There's no good way to take `myRawJSONExpression[0]` and refine it so that flow knows that it's an `ExpressionName`.